ee/src/main/java/org/jboss/as/ee/structure/InitializeInOrderProcessor.java

```diff
@@ -49,24 +50,45 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
         if (parent == null) {
             return;
         }
+
         final EarMetaData earConfig = parent.getAttachment(Attachments.EAR_METADATA);
-        if (earConfig != null) {
-            final boolean inOrder = earConfig.getInitializeInOrder();
-            if (inOrder && earConfig.getModules().size() > 1) {
 
-                final Map<String, DeploymentUnit> deploymentUnitMap = new HashMap<String, DeploymentUnit>();
-                for (final DeploymentUnit subDeployment : parent.getAttachment(org.jboss.as.server.deployment.Attachments.SUB_DEPLOYMENTS)) {
+        if (earConfig == null) {
+            return;
+        }
 
-                    final ResourceRoot deploymentRoot = subDeployment.getAttachment(org.jboss.as.server.deployment.Attachments.DEPLOYMENT_ROOT);
-                    final ModuleMetaData moduleMetaData = deploymentRoot.getAttachment(Attachments.MODULE_META_DATA);
-                    if (moduleMetaData != null) {
-                        deploymentUnitMap.put(moduleMetaData.getFileName(), subDeployment);
-                    }
+        final List<String> startupBeanModules = parent.getAttachmentList(org.jboss.as.ee.component.Attachments.DEPLOYMENT_STARTUP_MODULES);
+        final Map<String, DeploymentUnit> deploymentUnitMap = new HashMap<String, DeploymentUnit>();
+        final ResourceRoot deploymentRoot = deploymentUnit.getAttachment(org.jboss.as.server.deployment.Attachments.DEPLOYMENT_ROOT);
+        final ModuleMetaData thisModulesMetadata = deploymentRoot.getAttachment(Attachments.MODULE_META_DATA);
+
+        if((earConfig.getInitializeInOrder() || (!startupBeanModules.isEmpty())) && !startupBeanModules.contains(deploymentUnit.getName())) {
+            for (final DeploymentUnit subDeployment : parent.getAttachment(org.jboss.as.server.deployment.Attachments.SUB_DEPLOYMENTS)) {
+
+                final ResourceRoot subdeploymentRoot = subDeployment.getAttachment(org.jboss.as.server.deployment.Attachments.DEPLOYMENT_ROOT);
+                final ModuleMetaData moduleMetaData = subdeploymentRoot.getAttachment(Attachments.MODULE_META_DATA);
+                if (moduleMetaData != null) {
+                    deploymentUnitMap.put(moduleMetaData.getFileName(), subDeployment);
+                }
+            }
+        }
```

Comment:
For this block, the original code condition includes `earConfig.getModules().size() > 1`. Don't we still need this?